PURL-SPECIFICATION.rst

```diff
@@ -176,25 +178,106 @@ The rules for each component are:
+           .. code-block:: none
+
+               'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', '~'
+
+           - These 66 characters do not need to be percent-encoded.
+
+        2. **All US-ASCII characters defined as "sub-delims"**, a subset of
+           the "reserved" characters, in RFC 3986 Section 2.2
+           (https://datatracker.ietf.org/doc/html/rfc3986#section-2.2):
+
+           .. code-block:: none
+
+               '!', '$', '&', ''', '(', ')', '*', '+', ',', ';', '='
+
+           - The '&' MUST be percent-encoded to avoid being incorrectly parsed
+             as a separator between multiple key-value pairs. See "How to parse
+             a purl string in its components" ("Split the qualifiers on '&'.
+             Each part is a key=value pair").
+
+           - The other 10 characters do not need to be percent-encoded, including
+             the '=' -- the parser will not mistakenly treat a '=' in the value
+             as a separator because it splits each key-value pair just once,
+             from left-to-right, on the first '=' it encounters, and thus there
+             is no conflict:
+
+             .. code-block:: none
+
+                 - For each pair, split the key=value once from left on '=':
+                   - The key is the lowercase left side
+                   - The value is the percent-decoded right side
+
+        3. **Four additional US-ASCII characters** identified in the "query"
+           definition in RFC 3986 Section 3.4 (https://datatracker.ietf.org/doc/html/rfc3986#section-3.4)
+           and the "pchar" definition in RFC 3986 Appendix A (https://datatracker.ietf.org/doc/html/rfc3986#appendix-A):
+
+           .. code-block:: none
+
+               ':', '@', '/', '?'
+
+           - The '?' MUST be percent-encoded to avoid being incorrectly parsed
+             as a ``qualifiers`` separator -- in the right-to-left parsing
+             (see "How to parse a purl string in its components"), an unencoded
+             '?' in the ``value`` would be the first '?' encountered by the
+             parser and incorrectly treated as a ``qualifiers`` separator.
+
+           - The other three characters do not need to be percent-encoded.
+
+        4. **All other US-ASCII characters**.
+
+           .. code-block:: none
+
+               - 33 control characters (ASCII codes 0-31 and 127)
+
+               - the 14 US-ASCII characters not covered in the preceding groups of US-ASCII characters:
+
+                 ' ' [space], '"', '#', '%', '<', '>', '[', '\', ']', '^', '`', '{', '|', '}'
+
+           - Each of these 47 US-ASCII characters MUST be percent-encoded.
+
+        5. **Any character that is not a US-ASCII character**
+           (i.e., characters with ASCII code > 127 and non-ASCII characters).
+
+           - All such characters MUST be UTF-8 encoded and then percent-encoded.
+
+  - ``purl`` parsers MUST return an error when the ``key`` or ``value`` contains
+    a prohibited character.
```

Comment:
This is incorrect and incompatible with the preexisting spec.

In the unescaped form, no characters are prohibited, so you could have a valid PURL `pkg:generic/name?key=%00` and the parser must handle this (or maybe it returns an error because it's string type can't represent null characters).

In the escaped form, some characters are escaped to avoid problems unrelated to the URL. `http://example.com/?key=a "value"` and `http://example.com/?key=a%20%22value%22` are the same, but if you write `Go to http://example.com/?key=a "value"` or `<a href="http://example.com/?key=a "value"">` it doesn't work. At least I'm pretty sure why it's done. These characters have no meaning to the URL or PURL parsers so whether they are escaped or not doesn't make a difference to the parsing result.

Reply:
Seems to me that inclusion of a prohibited character could only be "normalized" by removing/discarding such a character, which does not sound to me like mere normalization.  I look to feedback from @pombredanne and others on this point.

Reply:
Maybe I'm misunderstanding. There are no characters that are prohibited in the unencoded form of a qualifier value. In the encoded form, there are characters that are listed above as requiring escaping that don't actually require escaping in order for the PURL to be parsed successfully. If it can be unambiguously parsed and the result can be formatted into a PURL with the same meaning, I don't think that should be a "MUST return an error" condition. The characters do not need to be removed or discarded.

Reply:
Hmmm.  I must admit that I don't yet have a good understanding of when a violation of the spec should be normalized (allowing the parsing process to continue) vs. treating a spec violation as invalid (i.e., `"is_invalid": true`, raising an error/exception and thus halting the parsing process).  

I'd be interested in reading any authorities/resources you could point me to on this important topic -- and in all events I'll happily defer to whatever approach you, @pombredanne, @jkowalleck, @mprpic and others in the community ultimately agree to.